workspaces/diff-engine/src/queries/shape.rs

```diff
@@ -60,21 +49,105 @@ impl<'a> ShapeQueries<'a> {
           _ => unreachable!("expected to be a core shape node"),
         };
-        ChoiceOutput {
-          parent_trail: shape_trail.clone(),
-          additional_components: vec![],
-          shape_id: shape_id.clone(),
-          core_shape_kind: core_shape_descriptor.kind.clone(),
-        }
+        let trails: Vec<ChoiceOutput> = match core_shape_descriptor.kind {
+          ShapeKind::UnknownKind => vec![],
+          ShapeKind::NullableKind => {
+            let nullable_parameter_id = core_shape_descriptor
+              .kind
+              .get_parameter_descriptor()
+              .unwrap()
+              .shape_parameter_id;
+            let item_shape_id =
+              self.resolve_parameter_to_shape(shape_id, &String::from(nullable_parameter_id));
+            let trail = shape_trail
+              .with_component(ShapeTrailPathComponent::NullableItemTrail {
+                shape_id: shape_id.clone(),
+                inner_shape_id: item_shape_id.clone(),
+              })
+              .with_component(ShapeTrailPathComponent::NullableItemTrail {
+                shape_id: shape_id.clone(),
+                inner_shape_id: item_shape_id.clone(),
+              });
+            vec![
+              vec![ChoiceOutput {
+                parent_trail: shape_trail.clone(),
+                additional_components: vec![ShapeTrailPathComponent::NullableTrail {
+                  shape_id: shape_id.clone(),
+                }],
+                shape_id: shape_id.clone(),
+                core_shape_kind: core_shape_descriptor.kind.clone(),
+              }],
+              self.list_trail_choices(&trail),
+            ]
+            .concat()
+          }
+          ShapeKind::OptionalKind => {
+            let optional_parameter_id = core_shape_descriptor
+              .kind
+              .get_parameter_descriptor()
+              .unwrap()
+              .shape_parameter_id;
+            let item_shape_id =
+              self.resolve_parameter_to_shape(shape_id, &String::from(optional_parameter_id));
+            let trail = shape_trail
+              .with_component(ShapeTrailPathComponent::OptionalItemTrail {
+                shape_id: shape_id.clone(),
+                inner_shape_id: item_shape_id.clone(),
+              })
+              .with_component(ShapeTrailPathComponent::OptionalItemTrail {
+                shape_id: shape_id.clone(),
+                inner_shape_id: item_shape_id.clone(),
+              });
+            vec![
+              vec![ChoiceOutput {
+                parent_trail: shape_trail.clone(),
+                additional_components: vec![ShapeTrailPathComponent::OptionalTrail {
+                  shape_id: shape_id.clone(),
+                }],
+                shape_id: shape_id.clone(),
+                core_shape_kind: core_shape_descriptor.kind.clone(),
+              }],
+              self.list_trail_choices(&trail),
+            ]
+            .concat()
```

Comment:
This messing about with vecs makes one wonder if `list_trail_choices` should (optionally) accept an existing vec to push onto, could prevent a bunch of allocations. Definitely an optimization without proof it's necessary.

Reply:
I think I can write it a better way now, will update!